### DISCUSSION TITLE
Conditional multiple response models

### DISCUSSION THREAD

**Discussion by ricardoorfao** [fastapi/fastapi#6819] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
import asyncio
from datetime import datetime
from typing import Union, Optional

import orjson
from fastapi import FastAPI
from hypercorn import Config
from hypercorn.asyncio import serve
from pydantic import BaseModel, EmailStr


def orjson_dumps(v, *, default):
    return orjson.dumps(v, default=default).decode()


class UserBase(BaseModel):
    email: Optional[EmailStr] = None
    confirmed_on: Optional[datetime] = None

    class Config:
        json_loads = orjson.loads
        json_dumps = orjson_dumps


class UserInDBBase(UserBase):
    id: int
    created_at: datetime

    class Config:
        orm_mode = True


class UserNormal(UserInDBBase):
    pass


class UserAdmin(UserInDBBase):
    updated_at: datetime


app = FastAPI(title="Test")


@app.get('/', response_model=Union[UserAdmin, UserNormal])
async def something(some: int):
    dict1 = {
        "id": 1,
        "email": "user1@domain.com",
        "confirmed_on": "2021-02-08 17:20:00",
        "created_at": "2021-01-17 15:00:00"
    }
    dict2 = {
        "id": 2,
        "email": "user2@domain.com",
        "confirmed_on": "2021-02-08 17:10:00",
        "created_at": "2021-01-17 15:00:00",
        "updated_at": "2021-01-17 15:00:00"
    }
    dict3 = {
        "id": 2,
        "email": "user2@domain.com",
        "confirmed_on": "2021-02-08 17:10:00",
        "created_at": "2021-01-17 15:00:00",
        "updated_at": None
    }

    if some == 1:
        return UserNormal(**dict1) # This works
    elif some == 2:
        return UserAdmin(**dict2) # This works
    else:
        return UserAdmin(**dict3) # Causes a crash


if __name__ == '__main__':
    asyncio.run(serve(app, Config()))  # noqa

```

#### Example 1

* Open the browser and call the endpoint `http://127.0.0.1:8000?some=1`.
* It returns a JSON with 
```json
{
    "email": "user1@domain.com",
    "confirmed_on": "2021-02-08T17:20:00",
    "id": 1,
    "created_at": "2021-01-17T15:00:00",
    "updated_at": null
}
```
* But I expected it to return 
```json
{
    "email": "user1@domain.com",
    "confirmed_on": "2021-02-08T17:20:00",
    "id": 1,
    "created_at": "2021-01-17T15:00:00"
}
```

#### Example 2

* Open the browser and call the endpoint `http://127.0.0.1:8000?some=2`.
* It returns a JSON with 
```json
{
    "email": "user2@domain.com",
    "confirmed_on": "2021-02-08T17:10:00",
    "id": 2,
    "created_at": "2021-01-17T15:00:00",
    "updated_at": "2021-01-17T15:00:00"
}
```
* This is exactly what I expect it to return.

I want some properties to be returned depending on the type of user requesting the information in the same endpoint. If the user is an administrator should receive the model that includes the `updated_at` property (it doesn't matter if the property is null). If the user isn't an administrator, he doesn't receive the property `updated_at`. Is there a way to achieve this?

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS
* FastAPI Version [e.g. 0.3.0]: 0.63.0
* Python version: 3.8.3

### Additional context

I simplified the example, but hopefully, you get the idea. I inspired myself on this [issue](https://github.com/tiangolo/fastapi/issues/1947).

**Comment by Kludex** [comment]
Yes.
 
```python
@app.get('/', response_model=Union[UserNormal, UserAdmin])
```

Order matters.

**Comment by ricardoorfao** [comment]
Thanks @Kludex. Changing the order results in the removal of `updated_at` for both models. I expect that `updated_at` is returned in the response when `some==2`.

**Comment by Kludex** [comment]
My suggestion was just wrong. You write it right. The most complex should come first, which is `Union[UserAdmin, UserNormal]` (https://fastapi.tiangolo.com/tutorial/extra-models/#union-or-anyof).

But the thing is that it matches the first, so there are several solutions here. The one that makes more sense to me, based on your case:

```python
class UserAdmin(UserInDBBase):
    updated_at: datetime
```

**Comment by ricardoorfao** [comment]
I see. I need it to be Optional, since it can be null.

**Comment by Kludex** [comment]
Then: https://fastapi.tiangolo.com/tutorial/response-model/#use-the-response_model_exclude_unset-parameter

The whole page is useful.

**Comment by ricardoorfao** [comment]
Yup, I read that. Not what I need. I'm probably not explaining this right...

If I add `response_model_exclude_unset=True` it will remove `updated_at` completely from UserAdmin model. I need UserNormal always without `updated_at` property and UserAdmin always with `updated_at` property, even when it's null. If I use
```python
class UserAdmin(UserInDBBase):
    updated_at: datetime
```
this causes a crash everythime I don't set the property. If I use:
```python
class UserAdmin(UserInDBBase):
    updated_at: Optional[datetime]
```
the property isn't returned in UserAdmin model. I added a case to the original question

**Comment by Kludex** [comment, marked as answer]
Have you tried this? :sweat_smile: 

```python
@app.get('/', response_model=Union[UserAdmin, UserNormal], response_model_exclude_unset=True)
```

**Comment by ricardoorfao** [comment]
It worked exactly as expected. Shame on me. Thanks a lot for the help @Kludex.

**Comment by Kludex** [comment]
You are welcome. Have a nice day! :)

**Comment by ricardoorfao** [comment]
For whoever needs a working solution for a similar problem (thanks to @Kludex):

```python
import asyncio
from datetime import datetime
from typing import Union, Optional

import orjson
from fastapi import FastAPI
from hypercorn import Config
from hypercorn.asyncio import serve
from pydantic import BaseModel, EmailStr


def orjson_dumps(v, *, default):
    return orjson.dumps(v, default=default).decode()


class UserBase(BaseModel):
    email: Optional[EmailStr] = None
    confirmed_on: Optional[datetime] = None

    class Config:
        json_loads = orjson.loads
        json_dumps = orjson_dumps


class UserInDBBase(UserBase):
    id: int
    created_at: datetime

    class Config:
        orm_mode = True


class UserNormal(UserInDBBase):
    pass


class UserAdmin(UserInDBBase):
    updated_at: Optional[datetime] # Optional to accept None values


app = FastAPI(title="Test")


@app.get('/', response_model=Union[UserAdmin, UserNormal], # order matters. The most complex should come first.
                      response_model_exclude_unset=True) # default values won't be included in the response
async def something(some: int):
    dict1 = {
        "id": 1,
        "email": "user1@domain.com",
        "confirmed_on": "2021-02-08 17:20:00",
        "created_at": "2021-01-17 15:00:00"
    }
    dict2 = {
        "id": 2,
        "email": "user2@domain.com",
        "confirmed_on": "2021-02-08 17:10:00",
        "created_at": "2021-01-17 15:00:00",
        "updated_at": "2021-01-17 15:00:00"
    }
    dict3 = {
        "id": 2,
        "email": "user2@domain.com",
        "confirmed_on": "2021-02-08 17:10:00",
        "created_at": "2021-01-17 15:00:00",
        "updated_at": None
    }

    if some == 1:
        return UserNormal(**dict1)
    elif some == 2:
        return UserAdmin(**dict2)
    else:
        return UserAdmin(**dict3)


if __name__ == '__main__':
    asyncio.run(serve(app, Config()))  # noqa

```

**Comment by tiangolo** [comment]
Thanks for the help here @Kludex ! 👏  🙇 

Thanks for reporting back and closing the issue @ricardoorfao  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
